src/cursor/cur_version.c

```diff
@@ -0,0 +1,316 @@
+    table_cursor = version_cursor->table_cursor;
+    table_cursor->set_key(table_cursor);
+}
+
+/*
+ * __curversion_get_key --
+ *     WT_CURSOR->get_key implementation for version cursors.
+ */
+static int
+__curversion_get_key(WT_CURSOR *cursor, ...)
+{
+    WT_CURSOR *table_cursor;
+    WT_CURSOR_VERSION *version_cursor;
+    WT_DECL_RET;
+    va_list ap;
+
+    version_cursor = (WT_CURSOR_VERSION *)cursor;
+    table_cursor = version_cursor->table_cursor;
+    va_start(ap, cursor);
+    WT_ERR(__wt_cursor_get_keyv(cursor, cursor->flags, ap));
+    WT_ERR(__wt_cursor_get_keyv(table_cursor, table_cursor->flags, ap));
+
+err:
+    va_end(ap);
+    return (ret);
+}
+
+/*
+ * __curversion_next --
+ *     WT_CURSOR->next method for version cursors. The next function will position the cursor on the
+ *     next update of the key it is positioned at. We traverse through updates on the update chain,
+ *     then the ondisk value, and finally from the history store.
+ */
+static int
+__curversion_next(WT_CURSOR *cursor)
+{
+    WT_CURSOR *hs_cursor, *table_cursor;
+    WT_CURSOR_BTREE *cbt;
+    WT_CURSOR_VERSION *version_cursor;
+    WT_DECL_RET;
+    WT_SESSION_IMPL *session;
+    WT_TIME_WINDOW tw;
+    WT_UPDATE *upd;
+
+    version_cursor = (WT_CURSOR_VERSION *)cursor;
+    hs_cursor = version_cursor->hs_cursor;
+    table_cursor = version_cursor->table_cursor;
+    cbt = (WT_CURSOR_BTREE *)table_cursor;
+    CURSOR_API_CALL(cursor, session, next, NULL);
+
+    /* The cursor should be positioned. */
+    if (!F_ISSET(cursor, WT_CURSTD_KEY_INT))
+        goto err;
+
+    upd = version_cursor->next_upd;
+
+    /* 
+     * If the update is a tombstone, we still want to record the stop information
+     * but we also need traverse to the next update to get the full value.
+     */
+    version_cursor->upd_txnid = upd->txnid;
+    version_cursor->upd_durable_stop_ts = upd->durable_ts;
+    version_cursor->upd_stop_ts = upd->start_ts;
+    if (upd->type == WT_UPDATE_TOMBSTONE)
+        upd = upd->next;
+
+    if (upd != NULL && !F_ISSET(version_cursor, WT_VERSION_CUR_UPDATE_EXHAUSTED)) {
+        /*
+         * Set the version cursor's key, which contains all the record metadata
+         * for that particular version of the update. 
+         */
+        __wt_cursor_set_key(cursor, upd->txnid, upd->durable_ts, upd->start_ts,
+          version_cursor->upd_txnid, version_cursor->upd_durable_stop_ts,
+          version_cursor->upd_stop_ts, upd->type, upd->prepare_state, upd->flags);
+
+        /*
+         * Set the version cursor's value to be the update's value. If the update
+         * is a modify, reconstruct the value.
+         */
+        if (upd->type != WT_UPDATE_MODIFY)
+            __wt_upd_value_assign(cbt->upd_value, upd);
+        else
+            __wt_modify_reconstruct_from_upd_list(session, cbt, upd, cbt->upd_value);
+
+        __wt_cursor_set_value(cursor, cbt->upd_value->buf);
+
+        upd = upd->next;
+        if (upd == NULL)
```

Comment:
If the last upd is aborted or a tombstone, we need to go to return the onpage value.